src/3d/qgs3dwindow.cpp

```diff
@@ -0,0 +1,124 @@
+#include <Qt3DCore/QEntity>
+#include <Qt3DExtras/QForwardRenderer>
+#include <Qt3DRender/QRenderSettings>
+#include <Qt3DRender/QRenderAspect>
+#include <Qt3DInput/QInputAspect>
+#include <Qt3DInput/QInputSettings>
+#include <Qt3DLogic/QLogicAspect>
+#include <Qt3DRender/QCamera>
+
+#include "qgs3dwindow.h"
+
+Qgs3DWindow::Qgs3DWindow()
+  : m_aspectEngine( new Qt3DCore::QAspectEngine )
+  , m_renderAspect( new Qt3DRender::QRenderAspect )
+  , m_inputAspect( new Qt3DInput::QInputAspect )
+  , m_logicAspect( new Qt3DLogic::QLogicAspect )
+  , m_renderSettings( new Qt3DRender::QRenderSettings )
+  , m_forwardRenderer( new Qt3DExtras::QForwardRenderer )
+  , m_defaultCamera( new Qt3DRender::QCamera )
+  , m_inputSettings( new Qt3DInput::QInputSettings )
+  , m_root( new Qt3DCore::QEntity )
+  , m_userRoot( nullptr )
+  , m_initialized( false )
+{
+
+  // register aspects
+  m_aspectEngine->registerAspect( m_renderAspect );
+  m_aspectEngine->registerAspect( m_inputAspect );
+  m_aspectEngine->registerAspect( m_logicAspect );
+
+  m_defaultCamera->setParent( m_root );
+  m_forwardRenderer->setCamera( m_defaultCamera );
+  m_forwardRenderer->setSurface( this );
+  m_renderSettings->setActiveFrameGraph( m_forwardRenderer );
+  m_inputSettings->setEventSource( this );
+}
+
+Qgs3DWindow::~Qgs3DWindow()
+{
+  delete m_aspectEngine;
+}
+
+void Qgs3DWindow::registerAspect( Qt3DCore::QAbstractAspect *aspect )
+{
+  Q_ASSERT( !isVisible() );
+  m_aspectEngine->registerAspect( aspect );
+}
+
+void Qgs3DWindow::registerAspect( const QString &name )
+{
+  Q_ASSERT( !isVisible() );
+  m_aspectEngine->registerAspect( name );
+}
+
+void Qgs3DWindow::setRootEntity( Qt3DCore::QEntity *root )
+{
+  if ( m_userRoot != root )
+  {
+    if ( m_userRoot != nullptr )
+      m_userRoot->setParent( static_cast<Qt3DCore::QNode *>( nullptr ) );
+    if ( root != nullptr )
+      root->setParent( m_root );
+    m_userRoot = root;
+  }
+}
+
+void Qgs3DWindow::setActiveFrameGraph( Qt3DRender::QFrameGraphNode *activeFrameGraph )
+{
+  m_renderSettings->setActiveFrameGraph( activeFrameGraph );
+}
+
+Qt3DRender::QFrameGraphNode *Qgs3DWindow::activeFrameGraph() const
+{
+  return m_renderSettings->activeFrameGraph();
+}
+
+Qt3DExtras::QForwardRenderer *Qgs3DWindow::defaultFrameGraph() const
+{
+  return m_forwardRenderer;
+}
+
+Qt3DRender::QCamera *Qgs3DWindow::camera() const
+{
+  return m_defaultCamera;
+}
+
+Qt3DRender::QRenderSettings *Qgs3DWindow::renderSettings() const
+{
+  return m_renderSettings;
+}
+
+void Qgs3DWindow::showEvent( QShowEvent *e )
+{
+  if ( !m_initialized )
+  {
+    m_root->addComponent( m_renderSettings );
+    m_root->addComponent( m_inputSettings );
+    m_aspectEngine->setRootEntity( Qt3DCore::QEntityPtr( m_root ) );
+
+    m_initialized = true;
+  }
+  QWindow::showEvent( e );
+}
+
+void Qgs3DWindow::resizeEvent( QResizeEvent * )
+{
+  m_defaultCamera->setAspectRatio( float( width() ) / std::max( 1.f, static_cast<float>( height() ) ) );
+}
```

Comment:
I did not yet figure out how to port this because it is making calls to QT private classes which are not exposed. But somehow, the 3d window still updates even without this method.

Reply:
In theory it should be possible to call into private Qt3D APIs from QGIS as well, although that's something to avoid...

Reply:
I did some further research on this, and I don't think that Qt3DWindow::event(QEvent *e) is needed, but I may be wrong. 

Looking at the events it is supposed to address:
•	QEvent::UpdateRequest – I did not find any situation when this event occurs
•	QEvent::Exposed – This event occurs after every resize of the 3d window for some reason. But in that case, the resizeEvent is triggered anyway.

While debugging this, I may have found another bug. I think that the 3dmapcanvas is being resized twice which leads to flickering on my system. I added a new commit which should address this.